# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [ ]:
import pandas as pd
import numpy as np

## Input Dataset

In [ ]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')

In [ ]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [ ]:
df

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0             1    15634602    France  Female   42       2       0.00   
1             2    15647311     Spain  Female   41       1   83807.86   
2             3    15619304    France  Female   42       8  159660.80   
3             4    15701354    France  Female   39       1       0.00   
4             5    15737888     Spain  Female   43       2  125510.82   
...         ...         ...       ...     ...  ...     ...        ...   
9995       9996    15606229    France    Male   39       5       0.00   
9996       9997    15569892    France    Male   35      10   57369.61   
9997       9998    15584532    France  Female   36       7       0.00   
9998       9999    15682355   Germany    Male   42       3   75075.31   
9999      10000    15628319    France  Female   28       4  130142.79   

      NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0                 1          1               1        101348.88       1  
1                 1          0               1        112542.58       0  
2                 3          1               0        113931.57       1  
3                 2          0               0         93826.63       0  
4                 1          1               1         79084.10       0  
...             ...        ...             ...              ...     ...  
9995              2          1               0         96270.64       0  
9996              1          1               1        101699.77       0  
9997              1          0               1         42085.58       1  
9998              2          1               0         92888.52       1  
9999              1          1               0         38190.78       0  

[10000 rows x 12 columns]

In [ ]:
df['Gender'].unique()

array(['Female', 'Male'], dtype=object)

In [ ]:
df['Geography'].unique()

array(['France', 'Spain', 'Germany'], dtype=object)

In [ ]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df = df[['Geography', 'Gender', 'Age', 'Tenure',
          'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
          'EstimatedSalary', 'Exited']].copy()

## Preprocessing

In [ ]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
df = pd.get_dummies(df, columns=['Geography', 'Gender'])

In [ ]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
X = df.drop('Exited', axis=1)  # Mengambil semua kolom kecuali 'Exited'
y = df['Exited']  # Mengambil kolom 'Exited' sebagai target (Y)

In [ ]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

#START CODE
scaler = MinMaxScaler()
X_transform = scaler.fit_transform(X)

In [ ]:
print("X (Fitur setelah scaling):")
print(X_transform)

print("\nY (Target):")
print(y)

X (Fitur setelah scaling):
[[0.32432432 0.2        0.         ... 0.         1.         0.        ]
 [0.31081081 0.1        0.33403148 ... 1.         1.         0.        ]
 [0.32432432 0.8        0.63635718 ... 0.         1.         0.        ]
 ...
 [0.24324324 0.7        0.         ... 0.         1.         0.        ]
 [0.32432432 0.3        0.29922631 ... 0.         0.         1.        ]
 [0.13513514 0.4        0.51870777 ... 0.         1.         0.        ]]

Y (Target):
0       1
1       0
2       1
3       0
4       0
       ..
9995    0
9996    0
9997    1
9998    1
9999    0
Name: Exited, Length: 10000, dtype: int64


In [ ]:
#Ini digunakan jika dilakukan scaler/Normalisas. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transform, columns = X.columns)
#X_transform = X.copy()

## Train-Test Split

In [ ]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

In [ ]:
X_train.shape, y_train.shape

((7500, 12), (7500,))

In [ ]:
print("Before Oversampling, counts of label '1':{}".format(sum(y_train ==1)))
print("Before Oversampling, counts of label '0':{}\n".format(sum(y_train ==0)))


from imblearn.over_sampling import SMOTE
sm = SMOTE(random_state =2)
x_train_res, y_train_res = sm.fit_resample(X_train, y_train.ravel())

print('After Oversamplng, the shape of train_x: {}'.format(x_train_res.shape))
print('After Oversamplng, the shape of train_y: {}\n'.format(x_train_res.shape))

print("After Oversamplng, counts of label '1': {}".format(sum(y_train_res == 1)))
print("After Oversamplng, counts of label '0' : {}".format(sum(y_train_res == 0)))

Before Oversampling, counts of label '1':1520
Before Oversampling, counts of label '0':5980

After Oversamplng, the shape of train_x: (11960, 12)
After Oversamplng, the shape of train_y: (11960, 12)

After Oversamplng, counts of label '1': 5980
After Oversamplng, counts of label '0' : 5980


# Modeling

## Model1
##KNN (K-Nearest Neighbors)

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!


K-Nearest Neighbors (KNN) adalah salah satu algoritma machine learning yang digunakan untuk tugas-tugas seperti klasifikasi dan regresi. Algoritma KNN bekerja berdasarkan konsep bahwa objek yang serupa cenderung berada dekat satu sama lain dalam ruang fitur. Ini adalah salah satu algoritma yang mudah dipahami dan dapat digunakan dalam berbagai kasus. KNN mengasumsikan bahwa jika memiliki dataset yang berisi poin-poin data dalam ruang fitur dapat diklasifikasikan objek baru dengan cara mencari k-nearest neighbors (k tetangga terdekat) dari objek tersebut dalam ruang fitur. Ini berarti objek baru akan diklasifikasikan berdasarkan mayoritas label k tetangga terdekatnya.

Nilai K adalah parameter yang dapat Anda tentukan, dan itu menentukan jumlah tetangga terdekat yang akan dipertimbangkan saat membuat prediksi. Anda dapat mencoba berbagai nilai K dan memilih yang memberikan hasil terbaik. Untuk menemukan tetangga terdekat, KNN menggunakan perhitungan jarak antara objek baru dan setiap poin data dalam dataset.

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
model = KNeighborsClassifier()
params = {'tol': [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}
grid_search = GridSearchCV(estimator=model, param_grid=params, cv=5)
model.fit(X_train,y_train)
y_pred = model.predict(X_test)

In [ ]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = model.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.85      0.95      0.90      1983
           1       0.63      0.36      0.46       517

    accuracy                           0.82      2500
   macro avg       0.74      0.65      0.68      2500
weighted avg       0.80      0.82      0.80      2500


[[1876  107]
 [ 333  184]]

0.824


## Model2
## Gradient Boosting
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!

Boosting adalah salah satu metode ensemble learning dalam machine learning yang bertujuan untuk meningkatkan kinerja model prediksi dengan menggabungkan beberapa model lemah menjadi satu model yang kuat. Ide dasar di balik Boosting adalah memberi bobot lebih pada data yang sulit dipecahkan oleh model lemah sebelumnya, sehingga model ensemble akhir lebih mampu menangani kasus-kasus sulit tersebut. Metode Boosting yang paling terkenal adalah AdaBoost (Adaptive Boosting) dan Gradient Boosting.

Pada model kedua ini saya menggunakan Gradient Boosting adalah salah satu metode ensemble learning yang digunakan dalam machine learning untuk meningkatkan kinerja model prediksi, terutama dalam masalah regresi dan klasifikasi. Ini adalah salah satu teknik yang sangat kuat dan sering digunakan dalam kompetisi data science. Seperti dalam metode ensemble lainnya, Gradient Boosting juga menggabungkan beberapa model lemah (biasanya pohon keputusan dangkal atau Decision Stump) menjadi satu model yang lebih kuat. Model lemah ini disebut "weak learner" karena mereka memiliki kinerja yang hanya sedikit di atas acak.

Proses Gradient Boosting adalah iteratif, yang berarti bahwa model ditingkatkan secara bertahap. Algoritma ini memulai dengan model lemah pertama dan kemudian mengubah model ini pada setiap iterasi berikutnya untuk meningkatkan performanya. Learning rate adalah parameter yang mengontrol sejauh mana model diperbarui pada setiap iterasi. Nilai learning rate yang lebih kecil cenderung menghasilkan konvergensi yang lebih lambat tetapi hasil yang lebih baik, sedangkan nilai yang lebih besar dapat menghasilkan konvergensi yang lebih cepat tetapi risiko melewati solusi yang optimal.

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
import xgboost as xgb
from sklearn.model_selection import GridSearchCV

# Membuat model XGBoost
model2 = xgb.XGBClassifier()

# Menentukan parameter grid yang akan diteliti
params = {
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 4, 5],
    'n_estimators': [100, 200, 300],
    'subsample': [0.8, 0.9, 1.0],
}

# Membuat objek GridSearchCV
grid = GridSearchCV(
    estimator=model2,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,
    cv=10
)

grid.fit(X_train,y_train)
# Menampilkan parameter terbaik
best_params = grid.best_params_
print("Parameter terbaik:", best_params)

Parameter terbaik: {'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 200, 'subsample': 0.8}


In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      0.96      0.92      1983
           1       0.77      0.50      0.61       517

    accuracy                           0.87      2500
   macro avg       0.83      0.73      0.77      2500
weighted avg       0.86      0.87      0.86      2500


[[1907   76]
 [ 256  261]]

0.8672


## Model3
## Support Vector Machine (SVM)
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

Model Support Vector Machine (SVM) adalah salah satu algoritma machine learning yang digunakan untuk klasifikasi dan regresi. Algoritma ini terutama digunakan dalam masalah klasifikasi, dan dapat digunakan dalam masalah dengan dua kelas (binary classification) maupun lebih dari dua kelas (multi-class classification).

SVM mencoba untuk membagi dataset menjadi dua kelas dengan mencari hyperplane terbaik yang memisahkan kelas-kelas ini dengan margin yang maksimal. Hyperplane ini merupakan batas keputusan yang digunakan untuk memprediksi kelas dari objek baru. Margin adalah jarak antara hyperplane dan titik-titik terdekat dari setiap kelas. Tujuan SVM adalah untuk mencari hyperplane yang memiliki margin maksimal, yang berarti jarak antara hyperplane dan titik-titik terdekat adalah sebesar mungkin. Hal ini juga berarti SVM mencari hyperplane yang memiliki generalisasi yang baik terhadap data yang belum pernah dilihat sebelumnya.
Support vectors adalah titik-titik data yang berada paling dekat dengan hyperplane, dan mereka adalah titik-titik yang paling penting dalam pemodelan SVM. Mereka digunakan untuk menentukan hyperplane yang optimal.

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
from sklearn.svm import SVC
model_svm = SVC()

model_svm.fit(X_train, y_train.ravel())
y_pred = model_svm.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print("Akurasi model SVM:", accuracy)



Akurasi model SVM: 0.8316


## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya
Kesimpulan yang dapat diambil dari hasil evaluasi tiga model yang saya gunakan di atas, model Gradient Boosting memiliki performa terbaik dengan accuracy score 0.8672.Kemudian yang kedua ada Model SVM dengan accuracy score 0.8316. Dan yang paling rendah berdasarkan evaluasi ini model ini adalah K-Nearest Neighbor.  